lib/jira/client/index.js

```diff
@@ -107,13 +130,24 @@ module.exports = async (id, installationId, jiraHost) => {
         delete: (repositoryId) => instance.delete(`/rest/devinfo/0.10/repository/${repositoryId}`, {
           fields: { _updateSequenceId: Date.now() }
         }),
-        update: (data, options) => instance.post('/rest/devinfo/0.10/bulk', {
-          preventTransitions: (options && options.preventTransitions) || false,
-          repositories: [data],
-          properties: {
-            installationId
+        update: (data, options) => {
+          // Deduplicate issue keys in commits and branches before request
+          if ("commits" in data) data.commits = dedupIssueKeys(data.commits)
+          if ("branches" in data) data.branches = dedupIssueKeys(data.branches)
+
+          if (issueKeySet(data.commits).length <= 400 && issueKeySet(data.branches).length <= 400) {
```

Comment:
![Screen Shot 2019-08-23 at 4 00 14 PM](https://user-images.githubusercontent.com/300976/63628020-257ec080-c5bf-11e9-9618-e178b045bc5f.png)

From [Atlassian API docs](https://developer.atlassian.com/cloud/jira/software/rest/#api-rest-devinfo-0-10-bulk-post). I believe the issue key limit per commit or branch is 100 and the commits array limit is 400.

---

We take the issues keys from every commit, put them in a set, then convert the set to an array. If that array contains for than 400 items, we bail out. Will this fail in cases when it should succeed?

For example, say we have 10 commits that each reference 40 unique issue keys. Each commit is under the issue key limit but when we add them together, there are 400 unique issue keys. Instead, does it make sense to get the max issue key count?

```javascript
const withinLimit = (resources) => {
  const issueKeyCounts = resource.map((resource) =>  resource.issueKeys.length)
  const maxIssueKeyCount = Math.max(...issueKeyCounts)

  return maxIssueKeyCount <= issueKeyLimit
}
```

Reply:
Another great catch! Just updated the logic here:https://github.com/integrations/jira/blob/ec19cc2d644e28e32a7eb5fa6b106562eeb065dd/lib/jira/client/index.js#L22-L27